rig-neo4j/src/lib.rs

```diff
@@ -0,0 +1,266 @@
+                        BoltType::Null(BoltNull) // Handle unexpected number type
+                    }
+                }
+                serde_json::Value::String(s) => BoltType::String(BoltString::new(&s)),
+                serde_json::Value::Array(arr) => BoltType::List(
+                    arr.iter()
+                        .map(|v| v.to_bolt_type())
+                        .collect::<Vec<BoltType>>()
+                        .into(),
+                ),
+                serde_json::Value::Object(obj) => {
+                    let mut bolt_map = BoltMap::new();
+                    for (k, v) in obj {
+                        bolt_map.put(BoltString::new(&k), v.to_bolt_type());
+                    }
+                    BoltType::Map(bolt_map)
+                }
+            },
+            Err(_) => {
+                println!("Couldn't serialize to JSON, will ignore.");
+                BoltType::Null(BoltNull) // Handle serialization error
+            }
+        }
+    }
+}
+
+impl Neo4jClient {
+    pub fn new(graph: Graph) -> Self {
+        Self { graph }
+    }
+
+    pub async fn connect(uri: &str, user: &str, password: &str) -> Result<Self, VectorStoreError> {
+        let graph = Graph::new(uri, user, password)
+            .await
+            .map_err(|e| VectorStoreError::DatastoreError(Box::new(e)))?;
+        Ok(Self { graph })
+    }
+
+    pub async fn from_config(config: Config) -> Result<Self, VectorStoreError> {
+        let graph = Graph::connect(config)
+            .await
+            .map_err(|e| VectorStoreError::DatastoreError(Box::new(e)))?;
+        Ok(Self { graph })
+    }
+
+    /// Returns a `Neo4jVectorIndex` that mirrors an existing Neo4j Vector Index.
+    ///
+    /// An index (of type "vector") of the same name as `index_name` must already exist for the Neo4j database.
+    /// See the Neo4j [documentation (Create vector index)](https://neo4j.com/docs/genai/tutorials/embeddings-vector-indexes/setup/vector-index/) for more information on creating indexes.
+    ///
+    /// ❗IMPORTANT: The index must be created with the same embedding model that will be used to query the index.
+    pub fn index<M: EmbeddingModel>(
+        &self,
+        model: M,
+        index_config: IndexConfig,
+        search_params: SearchParams,
+    ) -> Neo4jVectorIndex<M> {
+        Neo4jVectorIndex::new(self.graph.clone(), model, index_config, search_params)
+    }
+}
+
+#[allow(dead_code)]
+#[cfg(test)]
+mod tests {
+    use super::*;
+    use crate::vector_index::display::SearchResult;
+    use neo4rs::ConfigBuilder;
+    use rig::{
+        providers::openai::{Client, TEXT_EMBEDDING_ADA_002},
+        vector_store::VectorStoreIndex,
+    };
+    use serde::Deserialize;
+    use std::env;
+
+    const NEO4J_URI: &str = "neo4j+s://demo.neo4jlabs.com:7687";
+    const NEO4J_DB: &str = "recommendations";
+    const NEO4J_USERNAME: &str = "recommendations";
+    const NEO4J_PASSWORD: &str = "recommendations";
```

Comment:
Secrets in repo :grimacing:

Reply:
Not a secret! It's a demo database. 

https://github.com/neo4j-graph-examples/recommendations/tree/main?tab=readme-ov-file#setup